File: mobile/src/services/atlaspack/format.ts
/** Types miroir de services/api-geo/src/atlaspack/format.rs — ne pas diverger. */

export const FORMAT_VERSION = 1;
export const SCHEMA_VERSION = 1;

export const MANIFEST_FILE = 'manifest.json';
export const SIGNATURE_FILE = 'manifest.sig';
export const DATA_FILE = 'data.bin';
export const MBTILES_FILE = 'maps/offline.mbtiles';

export const RETURN_MANIFEST_FILE = 'manifest.json';
export const RETURN_MAC_FILE = 'manifest.mac';
export const RETURN_DATA_FILE = 'data.json';

export interface FileEntry {
  sha256: string;
  size_bytes: number;
}

export interface MapCoverage {
  zoom_min: number;
  zoom_max_requested: number;
  zoom_max_actual: number;
  tile_count: number;
  truncated: boolean;
  truncation_reason: string | null;
  bounds: [number, number, number, number];
}

export interface PackageManifest {
  format_version: number;
  schema_version: number;
  package_id: string;
  operator_user_id: string;
  operator_email: string;
  generated_at: string;
  expires_at: string;
  signing_key_id: string;
  mission_ids: string[];
  password_hash_algorithm: string;
  password_salt_b64: string;
  password_argon2_m_cost: number;
  password_argon2_t_cost: number;
  password_argon2_p_cost: number;
  password_hash_len: number;
  files: Record<string, FileEntry>;
  map_coverage: MapCoverage | null;
}

export interface OperatorIdentity {
  user_id: string;
  student_id: string | null;
  email: string;
  first_name: string | null;
  last_name: string | null;
  password_hash_phc: string;
}

export interface PackedPlannedPoint {
  id: string;
  numero: number;
  label: string | null;
  lat: number;
  lon: number;
  confirmed_sondage_id: string | null;
}

export interface PackedSondageMarker {
  id: string;
  code: string | null;
  longitude: number;
  latitude: number;
  status: string | null;
}

export interface PackedMission {
  id: string;
  code: string;
  title: string;
  theme: string;
  status: string;
  start_date: string | null;
  end_date: string | null;
  maille_id: string | null;
  maille_label: string | null;
  commune: string | null;
  region: string | null;
  expected_sondages: number;
  maille_geojson: unknown | null;
  bbox: [number, number, number, number] | null;
  center_lon: number | null;
  center_lat: number | null;
  tolerance_m: number;
  planned_points: PackedPlannedPoint[];
  existing_sondages: PackedSondageMarker[];
}

export interface OperatorPayload {
  schema_version: number;
  operator: OperatorIdentity;
  missions: PackedMission[];
  default_tolerance_m: number;
}

// ── .atlasreturn ────────────────────────────────────────────────────────

export interface ReturnManifest {
  format_version: number;
  export_id: string;
  package_id: string | null;
  operator_user_id: string;
  operator_email: string;
  generated_at: string;
  missions_count: number;
  sondages_count: number;
  essais_count: number;
  resultats_count: number;
  attachments_count: number;
  size_bytes: number;
  sha256: string;
  files: Record<string, FileEntry>;
}

export interface ReturnedSondage {
  id: string;
  mission_id: string;
  planned_point_id: string | null;
  code: string;
  longitude: number;
  latitude: number;
  location_accuracy_m: number | null;
  depth_m: number | null;
  layers_count: number | null;
  profile_description: string | null;
  notes: string | null;
  point_name: string | null;
  relocation_reason: string | null;
  created_at: string;
}

export interface ReturnedLabResult {
  id: string;
  mission_id: string;
  sondage_id: string;
  sample_code: string;
  depth_top_m: number;
  depth_bottom_m: number;
  sample: unknown;
  tests: unknown;
  status: string;
  created_at: string;
}

export interface ReturnedFieldLog {
  id: string;
  mission_id: string;
  log_type: string;
  content: string;
  longitude: number | null;
  latitude: number | null;
  created_at: string;
}

export interface ReturnedAttachment {
  id: string;
  sondage_id: string;
  mission_id: string | null;
  kind: string;
  file_name: string;
  content_type: string;
  caption: string | null;
  taken_at: string | null;
  archive_path: string;
  sha256: string;
  size_bytes: number;
}

export interface ReturnedAuditEvent {
  id: string;
  mission_id: string | null;
  event_type: string;
  occurred_at: string;
  object_type: string | null;
  object_id: string | null;
  old_values: unknown;
  new_values: unknown;
  metadata: unknown;
}

export interface ReturnData {
  schema_version: number;
  sondages: ReturnedSondage[];
  lab_results: ReturnedLabResult[];
  field_logs: ReturnedFieldLog[];
  attachments: ReturnedAttachment[];
  audit_events: ReturnedAuditEvent[];
}

File: services/api-geo/src/atlaspack/format.rs
//! Format `.atlaspack` / `.atlasreturn` — structures sérialisées, versionnées.
//!
//! Un `.atlaspack` est une archive ZIP contenant :
//!   - `manifest.json`   (clair, signé)
//!   - `manifest.sig`    (signature Ed25519 détachée de `manifest.json`)
//!   - `data.bin`        (JSON chiffré ChaCha20-Poly1305 : identité + missions)
//!   - `maps/offline.mbtiles` (optionnel, tuiles hors-ligne, non chiffré)
//!
//! `manifest.json` est un struct Rust à champs fixes : sa sérialisation JSON
//! est déterministe par construction (ordre des champs = ordre de
//! déclaration), donc `sha256(manifest.json bytes)` est stable — pas besoin
//! de tri de clés artificiel, sauf pour la map `files` (BTreeMap, déjà triée).

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Version du conteneur .atlaspack lui-même (structure de l'archive).
pub const FORMAT_VERSION: i32 = 1;
/// Version du schéma des données métier contenues dans `data.bin`.
pub const SCHEMA_VERSION: i32 = 1;

pub const MANIFEST_FILE: &str = "manifest.json";
pub const SIGNATURE_FILE: &str = "manifest.sig";
pub const DATA_FILE: &str = "data.bin";
pub const MBTILES_FILE: &str = "maps/offline.mbtiles";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileEntry {
    pub sha256: String,
    pub size_bytes: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MapCoverage {
    pub zoom_min: i32,
    pub zoom_max_requested: i32,
    pub zoom_max_actual: i32,
    pub tile_count: i64,
    pub truncated: bool,
    pub truncation_reason: Option<String>,
    pub bounds: [f64; 4], // [min_lon, min_lat, max_lon, max_lat]
}

/// Contenu de `manifest.json`. Signé dans son intégralité : toute
/// modification d'un octet invalide la signature.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PackageManifest {
    pub format_version: i32,
    pub schema_version: i32,
    pub package_id: Uuid,
    pub operator_user_id: Uuid,
    pub operator_email: String,
    /// RFC 3339
    pub generated_at: String,
    /// RFC 3339 — le mobile refuse tout import après cette date
    pub expires_at: String,
    pub signing_key_id: String,
    pub mission_ids: Vec<Uuid>,
    pub password_hash_algorithm: String,
    /// Salt + params Argon2id extraits du PHC string (dupliqués ici en clair
    /// pour que le mobile puisse dériver la clé de déchiffrement de `data.bin`
    /// AVANT de déchiffrer — ce ne sont pas des secrets : le salt et les
    /// coûts Argon2id ne compromettent rien sans le mot de passe réel).
    pub password_salt_b64: String,
    pub password_argon2_m_cost: u32,
    pub password_argon2_t_cost: u32,
    pub password_argon2_p_cost: u32,
    pub password_hash_len: u32,
    pub files: BTreeMap<String, FileEntry>,
    pub map_coverage: Option<MapCoverage>,
}

impl PackageManifest {
    /// Sérialisation canonique utilisée à la fois pour signer et pour
    /// vérifier — DOIT être strictement identique des deux côtés.
    pub fn canonical_bytes(&self) -> Vec<u8> {
        serde_json::to_vec(self).expect("PackageManifest est toujours sérialisable")
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OperatorIdentity {
    pub user_id: Uuid,
    pub student_id: Option<Uuid>,
    pub email: String,
    pub first_name: Option<String>,
    pub last_name: Option<String>,
    /// Chaîne PHC Argon2id complète, copiée telle quelle depuis
    /// `atlas.users.password_hash`. Le backend n'a jamais accès au mot de
    /// passe en clair, seulement à ce hash — c'est ce même hash que le
    /// mobile revérifie hors-ligne.
    pub password_hash_phc: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PackedPlannedPoint {
    pub id: Uuid,
    pub numero: i32,
    pub label: Option<String>,
    pub lat: f64,
    pub lon: f64,
    pub confirmed_sondage_id: Option<Uuid>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PackedSondageMarker {
    pub id: Uuid,
    pub code: Option<String>,
    pub longitude: f64,
    pub latitude: f64,
    pub status: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PackedMission {
    pub id: Uuid,
    pub code: String,
    pub title: String,
    pub theme: String,
    pub status: String,
    pub start_date: Option<String>,
    pub end_date: Option<String>,
    pub maille_id: Option<Uuid>,
    pub maille_label: Option<String>,
    pub commune: Option<String>,
    pub region: Option<String>,
    pub expected_sondages: i32,
    pub maille_geojson: Option<serde_json::Value>,
    /// [min_lon, min_lat, max_lon, max_lat] WGS84
    pub bbox: Option<[f64; 4]>,
    pub center_lon: Option<f64>,
    pub center_lat: Option<f64>,
    pub tolerance_m: i32,
    pub planned_points: Vec<PackedPlannedPoint>,
    pub existing_sondages: Vec<PackedSondageMarker>,
}

/// Contenu déchiffré de `data.bin`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OperatorPayload {
    pub schema_version: i32,
    pub operator: OperatorIdentity,
    pub missions: Vec<PackedMission>,
    pub default_tolerance_m: i32,
}

// ============================================================================
// .atlasreturn — export terrain réimporté au bureau
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReturnManifest {
    pub format_version: i32,
    pub export_id: Uuid,
    pub package_id: Option<Uuid>,
    pub operator_user_id: Uuid,
    pub operator_email: String,
    pub generated_at: String,
    pub missions_count: i32,
    pub sondages_count: i32,
    pub essais_count: i32,
    pub resultats_count: i32,
    pub attachments_count: i32,
    pub size_bytes: u64,
    pub sha256: String,
    pub files: BTreeMap<String, FileEntry>,
}

impl ReturnManifest {
    /// Octets signés par le HMAC (stocké séparément dans l'archive, fichier
    /// `manifest.mac`) — le manifeste tel quel, sous peine de circularité.
    pub fn canonical_bytes(&self) -> Vec<u8> {
        serde_json::to_vec(self).expect("ReturnManifest est toujours sérialisable")
    }

    /// `sha256` "racine" = sha256 de la concaténation triée des empreintes
    /// individuelles de `files` — recalculé et comparé au champ `sha256` du
    /// manifeste pour détecter une incohérence interne (ex: `files` modifié
    /// sans mettre à jour `sha256`), en plus de la vérification HMAC globale.
    pub fn compute_root_hash(&self) -> String {
        use sha2::Digest;
        let mut hashes: Vec<&str> = self.files.values().map(|f| f.sha256.as_str()).collect();
        hashes.sort_unstable();
        let joined = hashes.concat();
        let mut h = sha2::Sha256::new();
        h.update(joined.as_bytes());
        format!("{:x}", h.finalize())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReturnedSondage {
    pub id: Uuid,
    pub mission_id: Uuid,
    pub planned_point_id: Option<Uuid>,
    pub code: String,
    pub longitude: f64,
    pub latitude: f64,
    pub location_accuracy_m: Option<f32>,
    pub depth_m: Option<f64>,
    pub layers_count: Option<i32>,
    pub profile_description: Option<String>,
    pub notes: Option<String>,
    pub point_name: Option<String>,
    pub relocation_reason: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReturnedLabResult {
    pub id: Uuid,
    pub mission_id: Uuid,
    pub sondage_id: Uuid,
    pub sample_code: String,
    pub depth_top_m: f64,
    pub depth_bottom_m: f64,
    pub sample: serde_json::Value,
    pub tests: serde_json::Value,
    pub status: String,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReturnedFieldLog {
    pub id: Uuid,
    pub mission_id: Uuid,
    pub log_type: String,
    pub content: String,
    pub longitude: Option<f64>,
    pub latitude: Option<f64>,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReturnedAttachment {
    pub id: Uuid,
    pub sondage_id: Uuid,
    pub mission_id: Option<Uuid>,
    pub kind: String,
    pub file_name: String,
    pub content_type: String,
    pub caption: Option<String>,
    pub taken_at: Option<String>,
    /// Chemin de l'entrée binaire dans l'archive .atlasreturn (attachments/<id>.<ext>)
    pub archive_path: String,
    pub sha256: String,
    pub size_bytes: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReturnedAuditEvent {
    pub id: Uuid,
    pub mission_id: Option<Uuid>,
    pub event_type: String,
    pub occurred_at: String,
    pub object_type: Option<String>,
    pub object_id: Option<String>,
    pub old_values: Option<serde_json::Value>,
    pub new_values: Option<serde_json::Value>,
    pub metadata: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReturnData {
    pub schema_version: i32,
    pub sondages: Vec<ReturnedSondage>,
    pub lab_results: Vec<ReturnedLabResult>,
    pub field_logs: Vec<ReturnedFieldLog>,
    pub attachments: Vec<ReturnedAttachment>,
    pub audit_events: Vec<ReturnedAuditEvent>,
}
